rbs/AssertionsRewriter.cc

```diff
@@ -0,0 +1,565 @@
+    } else {
+        return nullopt;
+    }
+
+    return nullopt;
+}
+
+/**
+ * Parse the type parameters from the previous statement
+ *
+ * Given a case like this one:
+ *
+ *     #: [X] (X) -> void
+ *     def foo(x)
+ *       y = nil #: X?
+ *     end
+ *
+ * We need to be aware of the type parameter `X` so we can use it to resolve the type of `y`.
+ */
+vector<pair<core::LocOffsets, core::NameRef>> AssertionsRewriter::lastTypeParams() {
+    auto typeParams = vector<pair<core::LocOffsets, core::NameRef>>();
+
+    // Do we have a previous signature?
+    if (!lastSignature) {
+        return typeParams;
+    }
+
+    auto block = parser::cast_node<parser::Block>(lastSignature);
+    ENFORCE(block != nullptr);
+
+    // Does the sig contain a `type_parameters()` invocation?
+    auto send = parser::cast_node<parser::Send>(block->body.get());
+    while (send && send->method != core::Names::typeParameters()) {
+        send = parser::cast_node<parser::Send>(send->receiver.get());
+    }
+
+    if (send == nullptr) {
+        return typeParams;
+    }
+
+    // Collect the type parameters
+    for (auto &arg : send->args) {
+        auto sym = parser::cast_node<parser::Symbol>(arg.get());
+        if (sym == nullptr) {
+            continue;
+        }
+
+        typeParams.emplace_back(arg->loc, sym->val);
+    }
+
+    return typeParams;
+}
+
+void AssertionsRewriter::maybeSaveSignature(parser::Block *block) {
+    if (block->body == nullptr) {
+        return;
+    }
+
+    auto send = parser::cast_node<parser::Send>(block->send.get());
+    if (send == nullptr) {
+        return;
+    }
+
+    if (send->method != core::Names::sig()) {
+        return;
+    }
+
+    lastSignature = block;
+}
+
+unique_ptr<parser::Node> AssertionsRewriter::maybeInsertCast(unique_ptr<parser::Node> node) {
+    if (auto type = assertionForNode(node, core::LocOffsets::none())) {
+        return insertCast(move(node), move(type));
+    }
+
+    return node;
+}
+
+parser::NodeVec AssertionsRewriter::rewriteNodes(parser::NodeVec nodes) {
+    auto oldStmts = move(nodes);
+    auto newStmts = parser::NodeVec();
+
+    for (auto &node : oldStmts) {
+        node = rewriteNode(move(node));
+        newStmts.emplace_back(std::move(node));
+    }
+
+    return newStmts;
+}
+
+unique_ptr<parser::Node> AssertionsRewriter::rewriteBegin(unique_ptr<parser::Node> node) {
+    auto begin = parser::cast_node<parser::Begin>(node.get());
+    ENFORCE(begin != nullptr);
+
+    auto oldStmts = move(begin->stmts);
+    begin->stmts = parser::NodeVec();
+
+    for (auto &stmt : oldStmts) {
+        // stmt = maybeInsertCast(move(stmt));
+        begin->stmts.emplace_back(rewriteNode(move(stmt)));
+    }
+
+    return node;
+}
+
+unique_ptr<parser::Node> AssertionsRewriter::rewriteBody(unique_ptr<parser::Node> node) {
+    if (node == nullptr) {
+        return node;
+    }
+
+    if (auto begin = parser::cast_node<parser::Begin>(node.get())) {
```

Comment:
Not naming `begin` here should fix the emscripten build error.
```suggestion
    if (parser::cast_node<parser::Begin>(node.get()) != nullptr) {
```

Reply:
Thanks! I changed it for `isa_node` here and in the SigRewriter 👍